### ISSUE TITLE
We should support appliances in our analyses

### ISSUE THREAD

**Issue by sebbacon** [ebmdatalab/openprescribing#210]
NHS Prescription Services use a custom version of the BNF to code prescriptions. This is all the BNF, plus a bunch of extra codes ("pseudo-classifications") for things not in the BNF. Full details are [here](http://www.nhsbsa.nhs.uk/PrescriptionServices/Documents/BNF_Classification_Booklet_2009.pdf).  In particular, it includes pseudo-codes for appliances which in the BNF are listed in appendices rather than with fully coded entries.

The OpenPrescribing codebase assumes that there are chemicals and products associated with every prescribing line, and skips creating BNF chapters/sections/paras/products which start with the word `DUMMY` (which all the pseudo-codes do). This means it is impossible to analyses involving groups of appliances; you must list them one-by-one.

Increasingly we are getting requests for access to appliance data. There has been an increase in appliance prescribing across the board (see screenshot below); there are also, anecdotally, issues with waste (e.g. sterile packs of 20 pads being opened and destroyed to use a single pad).

So to take Hyabak eye drops, for example -- these are in the custom BNF only as two AMPs, viz. `Hyabak Sod Hyaluronate Eye Dps 10ml` and `Hyabak UD Sod Hyaluronate Eye Dps 0.4ml Ud P/F` but the products are `DUMMY PRODUCT 21300000108` and `DUMMY PRODUCT 21300000173` respectively (and the paragraph for both is `DUMMY PARAGRAPH 213000`). The lowest-level classification that both these presentations have in common is `DUMMY CHEMICAL SUBSTANCE 213000001`. 

So the challenge here is that most users would be interested in something that should be called `Hyabak Soc Hyaluronate` (rather than the two individual presentations), but it's actually called `DUMMY CHEMICAL SUBSTANCE 213000001` in the data.

To support appliance data, we would have to:
- Find a way to meaningfully label dummy chemical substances (either heuristically, or possibly through the D+DM mapping Lisa got hold of recently... needs checking to see if that has more useful classifications)
- Remove the code that skips dummy classifications
- Add code to hide dummy classifications in the user interface (I'm not sure without looking if there's any...)

@annapowellsmith, any other comments?

![download](https://cloud.githubusercontent.com/assets/211271/19038366/15336dce-8973-11e6-89ce-747c2e559ac0.png)

**Comment by annapowellsmith** [comment]
Cool - agree we should start supporting these!

We could skip the first bullet point - just import the `DUMMY` names as they are. Then hide them in the front-end, and just let expert users search by code. 

This would be easy enough on the back-end, and should only require a few tweaks in the front-end to allow users to search for `213000001` without showing `DUMMY` in the autocomplete.

**Comment by sebbacon** [comment]
The problem with the naive approach is that appliances tend to be at the AMPP level but people are interested in the AMP level (i.e. KoolBitz Bandages rather than KoolBitz Bandages 1m, 5m, and 10m as separate entities).  So expert users would end up having to do their own compound queries for all the KoolBitz bandages together.  

In other words, the question becomes: will it be useful enough for end users when they have to combine lots of AMPPs together in one query, or is that going to add too much mental overhead?

In turn, I suspect the answer to that question depends on exactly how many presentations of the same product there tend to be on average. If most appliances only come in 2 or 3 different forms, then it's not a biggie, but if lots of them come in 10s of forms, it may be.

**Comment by annapowellsmith** [comment]
Remind me what `AMPP` and `AMP` mean in BNF code terms?

**Comment by sebbacon** [comment]
Sorry, shouldn't have thrown `AMPP` in there, I used it wrongly. Sorry - explained in more detail below!  But for the purposes of this discussion, my point is people are unlikely to want to do a search for *Hyabak Sod Hyaluronate Eye Dps 10ml* **and** *Hyabak UD Sod Hyaluronate Eye Dps 0.4ml Ud P/F*. They will just want to search for *Hyabak Sod Hyaluronate*.

The degree to which this is a problem is the degree to which single "chemicals" (which in appliance-land might mean "dressings") have many different "presentations". Which we can find out with a bit of analysis.

----

To clear up AMPP etc: 

* **A** means **A**ctual, **V** means **V**irtual. Virtual means a generic name for a product (Paracetamol) versus Actual which is a brand sold by a manufacturer (e.g. Panadol).
* **MP** is **M**edicinal **P**roduct and should include the strength.
* The final (optional) **P** is **P**roduct and refers to an actual pack you can buy -- usually distinguished by pack size.

So Paracetamol 500mg tabs is a VMP. Paracetamol 500mg tabs 12 is a VMPP. Panadol 500mg tabs 12 is an AMPP. 

BNF data contains VMPs and AMPs but no pack sizes, which has caused some interesting problems in the Tariff work I've been doing in the background, because you can get one AMP in two very different pack sizes of very different prices.

```
Acarbose,0601023A0,Acarbose,0601023A0AA,Acarbose_Tab 50mg,0601023A0AAAAAA
^                     ^                     ^
|                     |                     |
chem                 prod                presentation
```

So the "presentation" in BNF-land is either an AMP or a VMP. *Not* as I said earlier an AMPP (or VMPP), When I said 

> The problem with the naive approach is that appliances tend to be at the AMPP level but people are interested in the AMP level 

It should have been 

> The problem with the naive approach is that appliances tend to be at the AMP level (presentation in BNF terms) but people are interested in the chemical or product level (in BNF terms)

I hope that clears it up :)

**Comment by annapowellsmith** [comment]
Thanks for the explanation - very clear!

So it will help to look at codes for a specific example. Following yours, let's look at these two products:

Hyabak Sod Hyaluronate Eye Dps 10ml (presentation code 21300000108, DUMMY PRODUCT 21300000108, DUMMY CHEMICAL SUBSTANCE 213000001)
Hyabak UD Sod Hyaluronate Eye Dps 0.4ml Ud P/F. (presentation code 21300000173, DUMMY PRODUCT 21300000173, DUMMY CHEMICAL SUBSTANCE 213000001)

So your concern is that people will want to search for the Hyabak Sod Hyaluronate product in general, but the only code-based search to do that is for `213000001`, which also includes lots of other products. I think that's correct?

I see the concern, but possibly YAGNI, and I don't think it should stop us importing them. I think we should import these products as they are, and then if people do get in touch and say that this is a problem for them, we should look at doing some name extraction, regex search, or whatever is needed.

**Comment by richiecroker** [comment]
The Hyabak is a good example of where we struggle with some of this.  Sodium Hyaluronate is used as a lubricant in the eye.  It also has a medicines code (1108010P0).  But, sometimes, for licensing reasons, manufacturers tend to have these classed as an appliance, rather than a medicine, so the usage is split over both areas - with the appliance codes NOT showing on openprescribing.

**Comment by richiecroker** [comment]
Also @sebbacon, interesting demonstration of the AMPP issue - but I thought that, for most drugs, although it used to be the case in the past that different pack sizes had different costs per pill, certainly for PartVIIIA of the drug tariff this has now been removed?

**Comment by sebbacon** [comment]
@annapowellsmith agreed possibly YAGNI, or at least, there's no real harm in adding these as nominators to our dataset. In the worst case they're not helpful because grouping them manually is too much effort, but there's already loads of chemicals, products etc to navigate in any case and we might find it's Good Enough.

@richiecroker so to summarise (for when we come back to this):

A single product can be listed as an appliance (in which case it has a pseudo-classification) or a medicine. For example, Hylo-Comod Eye Drops 0.1% are classified as a medicine, with a presentation code of `1108010P0BDAAAC` and a Chemical Substance code `1108010P0` (Sodium Hyaluronate). But Hyabek is classed as an appliance, with identical presentation and product code `21300000173` and a dummy chemical substance code `213000001`. 

So there are actually two reasons why appliances are problematic for analyses, both because denominators are hard to make meaningful:

1. When they are in pseudoclassifications the "chemical" or "subparagraph" doesn't have a meaningful name
2. The products are sometimes (often?) classified as medicines, meaning the denominators are (very or somewhat) likely to be incomplete.

I've had a quick look at the DM+D data and unfortunately the only other classification system it provides (ATC) does not have classifications for appliances.

The conclusion: if we want to support rich analyses of appliances, we will have to do extra work to devise useful alternative denominators; develop names for the dummy sections / chemicals; or encourage NHSBSA to do so.

There are 21080 DUMMY presentations in 682 paragraphs, 763 sub-paragraphs, and 1063 dummy chemicals. Examples of dummy chemicals with many (~100) presentations are:

* Hernia support belts (come in many sizes)
* Compression stockings
* Incontinence appliances

So this does seem like something we could develop quite easily with a medical librarian in a few weeks, if we wanted to look for funding for it.

**Comment by annapowellsmith** [comment]
Good discussion, thanks all!

I think the consensus is that the short-term task is to add them in the naive way, then return to this discussion if we find that users want to search in different ways.

**Comment by HelenCEBM** [comment]
User request received on this topic:

> I'm struggling to break down Ch21.4 'contraceptive devices'. At the moment, IUDs (copper coils) are in the same group as diaphragms and caps, and there doesn't appear to be a way to divide them up.

> I can do it with the NHS Business Services Authority raw data (by searching for 'iucd' or a BNF code that begins '2104000002')

In this case a workaround is selecting all ~28 products of interest. 

However, for advanced users it could be useful to offer these numerical searches even where names for the pseudo-chemicals don't exist.

**Comment by sebbacon** [comment]
The UI challenge side of this would be addressed by #953
